File: slick-grid-demo/app/src/lib/quickSort.ts
export type Cmp<T> = (a: T, b: T) => -1 | 0 | 1;

function swap<T>(arr: Array<T>, i: number, j: number): void {
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

function partition<T>(arr: Array<T>, low: number, high: number, cmp: Cmp<T>): number {
    let pivot = arr[high];

    let i = low - 1;
    for (let j = low; j <= high; j++) {
        if (cmp(arr[j], pivot) <= 0) {
            i += 1;
            swap(arr, i, j);
        }
    }
    return i;
}

function quickSortInner<T>(arr: Array<T>, low: number, high: number, cmp: Cmp<T>): void {
    if (low >= 0 && high >= 0 && low < high) {
        let p = partition(arr, low, high, cmp);
        quickSortInner(arr, low, p - 1, cmp);
        quickSortInner(arr, p + 1, high, cmp);
    }
}

function quickSort<T>(arr: Array<T>, cmp: Cmp<T>): Array<T> {
    quickSortInner(arr, 0, arr.length - 1, cmp);
    return arr;
}

export default quickSort;

File: websys-react-demo/wasm/src/quick_sort.rs
use core::cmp::Ordering;

fn partition<T>(arr: &mut Vec<T>, low: isize, high: isize, cmp: fn(&T, &T) -> Ordering) -> isize {
    let pivot = high as usize;

    let mut i = low - 1;

    for j in low..(high + 1) {
        let ord = cmp(&arr[j as usize], &arr[pivot]);
        if ord == Ordering::Less || ord == Ordering::Equal {
            i += 1;
            arr.swap(i as usize, j as usize);
        }
    }

    i
}

fn quick_sort_inner<T>(arr: &mut Vec<T>, low: isize, high: isize, cmp: fn(&T, &T) -> Ordering) {
    if low >= 0 && high >= 0 && low < high {
        let p = partition(arr, low, high, cmp);
        quick_sort_inner(arr, low, p - 1, cmp);
        quick_sort_inner(arr, p + 1, high, cmp)
    }
}

pub fn quick_sort<T>(arr: &mut Vec<T>, cmp: fn(&T, &T) -> Ordering) {
    let length = arr.len();
    quick_sort_inner(arr, 0, (length - 1) as isize, cmp);
}
